File: src/main/java/com/todoroo/astrid/dao/TaskListMetadataDao.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.dao;

import com.todoroo.andlib.data.Property;
import com.todoroo.andlib.sql.Criterion;
import com.todoroo.andlib.sql.Query;
import com.todoroo.astrid.data.TagData;
import com.todoroo.astrid.data.TaskListMetadata;

import javax.inject.Inject;

/**
 * Data Access layer for {@link TagData}-related operations.
 *
 * @author Tim Su <tim@todoroo.com>
 */
public class TaskListMetadataDao {

    private final RemoteModelDao<TaskListMetadata> dao;

    @Inject
    public TaskListMetadataDao(Database database) {
        dao = new RemoteModelDao<>(database, TaskListMetadata.class);
    }

    public TaskListMetadata fetchByTagId(String tagUuid, Property<?>... properties) {
        return dao.getFirst(Query.select(properties).where(Criterion.or(TaskListMetadata.TAG_UUID.eq(tagUuid),
                TaskListMetadata.FILTER.eq(tagUuid))));
    }

    public void createNew(TaskListMetadata taskListMetadata) {
        dao.createNew(taskListMetadata);
    }

    public void saveExisting(TaskListMetadata list) {
        dao.saveExisting(list);
    }
}

File: data/src/main/kotlin/org/tasks/data/dao/TaskListMetadataDao.kt
package org.tasks.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import org.tasks.data.entity.TaskListMetadata

@Dao
abstract class TaskListMetadataDao {
    @Query("SELECT * from task_list_metadata where tag_uuid = :tagUuid OR filter = :tagUuid LIMIT 1")
    abstract suspend fun fetchByTagOrFilter(tagUuid: String): TaskListMetadata?

    @Query("SELECT * FROM task_list_metadata")
    abstract suspend fun getAll(): List<TaskListMetadata>

    @Update
    abstract suspend fun update(taskListMetadata: TaskListMetadata)

    @Insert
    abstract suspend fun insert(taskListMetadata: TaskListMetadata): Long

    suspend fun createNew(taskListMetadata: TaskListMetadata) {
        taskListMetadata.id = insert(taskListMetadata)
    }
}